src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -528,3 +570,54 @@ object IndexLogEntry {
 
   def schemaString(schema: StructType): String = schema.json
 }
+
+/**
+ * Provides functionality to generate unique file ids for files.
+ */
+class FileIdTracker {
+  private var maxId: Long = -1L
+
+  // Combination of file properties, used as key, to identify a
+  // unique file for which an id is generated.
+  type key = (
+    String, // Full path.
+      Long, // Size.
+      Long  // Modified time.
+    )
+  private val fileToIdMap: mutable.HashMap[key, Long] = mutable.HashMap()
+
+  def getMaxFileId: Long = maxId
+
+  def getFileToIdMap: HashMap[key, Long] = fileToIdMap
+
+  def getFileId(path: String, size: Long, modifiedTime: Long): Option[Long] =
+    fileToIdMap.get((path, size, modifiedTime))
+
+  def setSizeHint(size: Int): Unit = fileToIdMap.sizeHint(size)
+
+  // FileInfo's 'name' contains the full path to the file.
+  def addFileIds(files: Set[FileInfo]): Unit = {
+    setSizeHint(files.size)
+    files.foreach { f =>
+      fileToIdMap.put((f.name, f.size, f.modifiedTime), f.id)
```

Comment:
Don't you need check the key doesn't already exist since we are trying to respect the existing id?

Reply:
Check added.